vnext/Desktop/WinHTTPWebSocketResource.cpp

```diff
@@ -0,0 +1,447 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// clang-format off
+#include "pch.h"
+
+#include <Unicode.h>
+#include <Utils.h>
+#include "WinHTTPWebSocketResource.h"
+
+// Windows API
+#include <comutil.h>
+
+using Microsoft::Common::Unicode::Utf8ToUtf16;
+
+using std::function;
+using std::size_t;
+using std::string;
+using std::wstring;
+
+namespace Microsoft::React
+{
+
+WinHTTPWebSocketResource::WinHTTPWebSocketResource(const string& urlString)
+  : m_openSessionFlags{ WINHTTP_FLAG_ASYNC }
+  , m_sessionHandle{ NULL }
+  , m_connectionHandle{ NULL }
+  , m_requestHandle{ NULL }
+  , m_errorStatus{ ERROR_SUCCESS }
+{
+#if 0 // WinHttpCrackUrl does not accept non-HTTP schemes 🙄
+  BSTR bstr = _com_util::ConvertStringToBSTR(urlString.c_str());
+  LPWSTR lpwstr = bstr;
+  SysFreeString(bstr);
+
+  LPURL_COMPONENTS url{};
+  auto parseResult = WinHttpCrackUrl(lpwstr, static_cast<DWORD>(urlString.length()), /*flags*/ 0, url);
+  if (!parseResult)
+    throw std::exception("Could not parse URL.");
+
+  if (wcscmp(L"wss", url->lpszScheme) == 0)
+  {
+    m_openSessionFlags |= WINHTTP_FLAG_SECURE;
+  }
+  else if (wcscmp(L"ws", url->lpszScheme) != 0)
+  {
+    throw std::exception(Common::Unicode::Utf16ToUtf8(
+      std::wstring(L"Incorrect URL scheme: ") + std::wstring(url->lpszScheme)
+    ).c_str());
```

Comment:
>.c_str() [](start = 5, length = 8)

are we allowed to throw a pointer to a value on the stack? object underneath should get destroyed during exception stack unroll

Reply:
this is fine because you're taking the c_str or a temporary, passing that to Utf16ToUtf8 which will grab that and construct a string object, then passing that object to the invalid_argument constructor, which will make a copy. The lifetime of the temporaries extends to the semicolon ;

---
In reply to: [393885907](https://github.com/microsoft/react-native-windows/pull/4330#discussion_r393885907) [](ancestors = 393885907)